fix: redirect ouput and errors to appropriate writer

KuthumiPepple opened the pull request:
<!-- Description -->


Closes #2773 
fmt.Print...() methods have been removed where applicable and outputs and error messages are now being written to the appropriate Writers. @astefanutti please review

<!--
Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". 

You can (optionally) mark this PR with labels "kind/bug" or "kind/feature" to make sure
the text is added to the right section of the release notes. 
-->

**Release Note**
```release-note
NONE
```

tadayosi commented:
There are linter validation errors. Please fix them.

astefanutti commented on e2e/support/test_support.go:
```diff
@@ -1719,9 +1719,9 @@ func UserCleanup() {
 		command.Stdout = os.Stdout
 		if err := command.Run(); err != nil {
-			fmt.Printf("An error occurred during user cleanup command execution: %v\n", err)
+			fmt.Fprintf(command.Stderr, "An error occurred during user cleanup command execution: %v\n", err)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Ok. I'd make the change

KuthumiPepple replied:
Looking at this again, `Fprintln` doesn't accept a format string. Using `Fprintln` would mean doing away with the formatting. Isn't that inportant?

astefanutti replied:
`Fprintln` operands are appended with white spaces, e.g.:

```golang
fmt.Fprintf(command.Stderr, "An error occurred during user cleanup command execution: %v\n", err)
```

Can be turned into:

```golang
fmt.Fprintln(command.Stderr, "An error occurred during user cleanup command execution:", err.Error())
```

I think for all the statements where formatting is actually used for appending strings, then `Fprintln` is more appropriate.

KuthumiPepple replied:
Ok thank you. I'll use that

astefanutti commented on e2e/support/test_support.go:
```diff
@@ -1719,9 +1719,9 @@ func UserCleanup() {
+			fmt.Fprintf(command.Stderr, "An error occurred during user cleanup command execution: %v\n", err)
 		} else {
-			fmt.Printf("User cleanup command completed successfully\n")
+			fmt.Fprintf(command.Stdout, "User cleanup command completed successfully\n")
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/bind.go:
```diff
@@ -219,9 +219,9 @@ func (o *bindCmdOptions) run(cmd *cobra.Command, args []string) error {
 
 	if !existed {
-		fmt.Printf("kamelet binding \"%s\" created\n", name)
+		fmt.Fprintf(cmd.OutOrStdout(), "kamelet binding \"%s\" created\n", name)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/bind.go:
```diff
@@ -219,9 +219,9 @@ func (o *bindCmdOptions) run(cmd *cobra.Command, args []string) error {
+		fmt.Fprintf(cmd.OutOrStdout(), "kamelet binding \"%s\" created\n", name)
 	} else {
-		fmt.Printf("kamelet binding \"%s\" updated\n", name)
+		fmt.Fprintf(cmd.OutOrStdout(), "kamelet binding \"%s\" updated\n", name)
```
Could be replaced with `Fprintln`.

astefanutti commented on pkg/cmd/run.go:
```diff
@@ -615,10 +615,10 @@ func (o *runCmdOptions) createOrUpdateIntegration(cmd *cobra.Command, c client.C
 	if existing == nil {
 		err = c.Create(o.Context, integration)
-		fmt.Printf("Integration \"%s\" created\n", name)
+		fmt.Fprintf(cmd.OutOrStdout(), "Integration \"%s\" created\n", name)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/run.go:
```diff
@@ -615,10 +615,10 @@ func (o *runCmdOptions) createOrUpdateIntegration(cmd *cobra.Command, c client.C
 	} else {
 		err = c.Patch(o.Context, integration, ctrl.MergeFromWithOptions(existing, ctrl.MergeFromWithOptimisticLock{}))
-		fmt.Printf("Integration \"%s\" updated\n", name)
+		fmt.Fprintf(cmd.OutOrStdout(), "Integration \"%s\" updated\n", name)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/run_help.go:
```diff
@@ -43,20 +44,20 @@ func hashFrom(contents ...[]byte) string {
 		cm := kubernetes.LookupConfigmap(ctx, c, integration.Namespace, config.Name())
 		if cm == nil {
-			fmt.Printf("Warn: %s Configmap not found in %s namespace, make sure to provide it before the Integration can run\n",
+			fmt.Fprintf(cmd.ErrOrStderr(), "Warn: %s Configmap not found in %s namespace, make sure to provide it before the Integration can run\n",
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/run_help.go:
```diff
@@ -43,20 +44,20 @@ func hashFrom(contents ...[]byte) string {
 		secret := kubernetes.LookupSecret(ctx, c, integration.Namespace, config.Name())
 		if secret == nil {
-			fmt.Printf("Warn: %s Secret not found in %s namespace, make sure to provide it before the Integration can run\n",
+			fmt.Fprintf(cmd.ErrOrStderr(), "Warn: %s Secret not found in %s namespace, make sure to provide it before the Integration can run\n",
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -186,21 +186,21 @@ func (o *uninstallCmdOptions) uninstallOperator(ctx context.Context, c client.Cl
 			return err
 		}
-		fmt.Printf("Camel K Custom Resource Definitions removed from cluster\n")
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K Custom Resource Definitions removed from cluster\n")
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 			return err
 		}
-		fmt.Printf("Camel K Role Bindings removed from namespace %s\n", o.Namespace)
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K Role Bindings removed from namespace %s\n", o.Namespace)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 			return err
 		}
-		fmt.Printf("Camel K Roles removed from namespace %s\n", o.Namespace)
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K Roles removed from namespace %s\n", o.Namespace)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 			return err
 		}
-		fmt.Printf("Camel K Service Accounts removed from namespace %s\n", o.Namespace)
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K Service Accounts removed from namespace %s\n", o.Namespace)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 			return err
 		}
-		fmt.Printf("Camel K Config Maps removed from namespace %s\n", o.Namespace)
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K Config Maps removed from namespace %s\n", o.Namespace)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 			return err
 		}
-		fmt.Printf("Camel K Registry Secret removed from namespace %s\n", o.Namespace)
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K Registry Secret removed from namespace %s\n", o.Namespace)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/util_containerization.go:
```diff
@@ -192,7 +196,7 @@ func createAndBuildIntegrationImage(ctx context.Context, containerRegistry strin
 
 	// Output executed command.
-	fmt.Printf("Executing: %s\n", strings.Join(cmd.Args, " "))
+	fmt.Fprintf(cmd.Stdout, "Executing: %s\n", strings.Join(cmd.Args, " "))
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
Done

astefanutti commented on pkg/cmd/util_containerization.go:
```diff
@@ -94,8 +94,12 @@ func createAndBuildBaseImage(ctx context.Context) error {
+
 	// Output executed command.
-	fmt.Printf("Executing: %s\n", strings.Join(cmd.Args, " "))
+	fmt.Fprintf(cmd.Stdout, "Executing: %s\n", strings.Join(cmd.Args, " "))
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
done

astefanutti commented on pkg/cmd/util_commands.go:
```diff
@@ -108,7 +108,7 @@ func RunLocalIntegrationRunCommand(ctx context.Context, properties []string, dep
 
 	// Output command we are about to run.
-	fmt.Printf("Executing: %s\n", strings.Join(cmd.Args, " "))
+	fmt.Fprintf(cmd.Stdout, "Executing: %s\n", strings.Join(cmd.Args, " "))
```
Could be replaced with `Fprintln`.

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 			return err
 		}
-		fmt.Printf("Camel K platform Kamelets removed from namespace %s\n", o.Namespace)
+		fmt.Fprintf(cmd.OutOrStdout(), "Camel K platform Kamelets removed from namespace %s\n", o.Namespace)
```
Could be replaced with `Fprintln`.

astefanutti commented on pkg/cmd/util_containerization.go:
```diff
@@ -225,7 +229,7 @@ func runIntegrationImage(ctx context.Context, image string, stdout, stderr io.Wr
 
 	// Output executed command.
-	fmt.Printf("Executing: %s\n", strings.Join(cmd.Args, " "))
+	fmt.Fprintf(cmd.Stdout, "Executing: %s\n", strings.Join(cmd.Args, " "))
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
done

astefanutti commented on pkg/cmd/util_dependencies.go:
```diff
@@ -230,37 +231,37 @@ func createCamelCatalog(ctx context.Context) (*camel.RuntimeCatalog, error) {
+		fmt.Fprintln(cmd.OutOrStdout(), "dependencies:")
 		for _, dep := range dependencies {
-			fmt.Printf("%v\n", dep)
+			fmt.Fprintf(cmd.OutOrStdout(), "%v\n", dep)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
done

astefanutti commented on pkg/cmd/version.go:
```diff
@@ -171,15 +171,15 @@ func operatorVersion(ctx context.Context, c client.Client, namespace string) (st
 	a, err := semver.NewVersion(aVersion)
 	if err != nil {
-		fmt.Printf("Could not parse %s (error: %s)\n", a, err)
+		fmt.Fprintf(cmd.ErrOrStderr(), "Could not parse %s (error: %s)\n", a, err)
```
Could be replaced with `Fprintln`.

astefanutti commented on pkg/cmd/version.go:
```diff
@@ -171,15 +171,15 @@ func operatorVersion(ctx context.Context, c client.Client, namespace string) (st
 	b, err := semver.NewVersion(bVersion)
 	if err != nil {
-		fmt.Printf("Could not parse %s (error: %s)\n", b, err)
+		fmt.Fprintf(cmd.ErrOrStderr(), "Could not parse %s (error: %s)\n", b, err)
```
Could be replaced with `Fprintln`.

KuthumiPepple replied:
done

astefanutti commented on e2e/support/test_support.go:
```diff
@@ -1719,9 +1719,9 @@ func UserCleanup() {
 		command.Stdout = os.Stdout
 		if err := command.Run(); err != nil {
-			fmt.Printf("An error occurred during user cleanup command execution: %v\n", err)
+			fmt.Fprintf(command.Stderr, "An error occurred during user cleanup command execution: %v\n", err)
```
In that case, I think it's more appropriate to use the test logger with `t.Logf`.

KuthumiPepple replied:
Yes that makes sense. I'd use the test logger

astefanutti commented on e2e/support/test_support.go:
```diff
@@ -1719,9 +1719,9 @@ func UserCleanup() {
+			fmt.Fprintf(command.Stderr, "An error occurred during user cleanup command execution: %v\n", err)
 		} else {
-			fmt.Printf("User cleanup command completed successfully\n")
+			fmt.Fprintf(command.Stdout, "User cleanup command completed successfully\n")
```
In that case, I think it's more appropriate to use the test logger with `t.Logf`.

squakez commented on pkg/cmd/delete.go:
```diff
@@ -139,14 +139,14 @@ func getIntegration(ctx context.Context, c client.Client, name string, namespace
 	if deleted {
 		// Deleting KameletBinding will automatically clean up the integration
-		fmt.Println("KameletBinding " + binding + " deleted")
+		fmt.Fprintln(cmd.OutOrStdout(),"KameletBinding " + binding + " deleted")
```
Missing space before the "KameletBinding" parameter.

KuthumiPepple replied:
fixed

squakez commented on pkg/cmd/rebuild.go:
```diff
@@ -69,7 +69,7 @@ func (o *rebuildCmdOptions) rebuild(_ *cobra.Command, args []string) error {
 	}
 
-	fmt.Printf("%d integrations have been rebuilt\n", len(integrations))
+	fmt.Fprintf(cmd.OutOrStdout(),"%d integrations have been rebuilt\n", len(integrations))
```
Another missing space. Likely you want to run `gofmt` to fix all formatting errors.

squakez commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 }
 
-func (o *uninstallCmdOptions) uninstallNamespaceRoles(ctx context.Context, c client.Client) error {
+func (o *uninstallCmdOptions) uninstallNamespaceRoles(cmd *cobra.Command, ctx context.Context, c client.Client) error {
```
Lint error, please fix accordingly.

squakez commented on pkg/cmd/util_dependencies.go:
```diff
@@ -230,37 +231,37 @@ func createCamelCatalog(ctx context.Context) (*camel.RuntimeCatalog, error) {
 			return err
 		}
-		fmt.Print(string(data))
+		fmt.Fprint(cmd.OutOrStdout(),string(data))
```
Lint error, please fix accordingly.

KuthumiPepple replied:
fixed

KuthumiPepple commented on pkg/cmd/delete.go:
```diff
@@ -139,14 +139,14 @@ func getIntegration(ctx context.Context, c client.Client, name string, namespace
 }
 
-func deleteIntegration(ctx context.Context, c client.Client, integration *v1.Integration) error {
+func deleteIntegration(cmd *cobra.Command, ctx context.Context, c client.Client, integration *v1.Integration) error {
```
done

astefanutti commented on pkg/cmd/debug.go:
```diff
@@ -100,15 +100,15 @@ func (o *debugCmdOptions) run(cmd *cobra.Command, args []string) error {
 			return
 		}
-		fmt.Printf("Disabling debug mode on integration %q\n", name)
+		fmt.Fprintf(cmd.OutOrStdout(), "Disabling debug mode on integration %q\n", name)
```
`Fprintln` can be used here also.

KuthumiPepple replied:
I did miss some of those Fprintf statements. I'll try to get them all this time. If I may ask, given the purpose of my next commit, would it not be better if do a commit amend and force push rather than do a new commit?

astefanutti replied:
No worries. I agree, amending and force push is a good idea.

astefanutti commented on pkg/cmd/describe_kamelet.go:
```diff
@@ -82,12 +82,12 @@ func (command *describeKameletCommandOptions) run(cmd *cobra.Command, args []str
 		}
 	} else {
-		fmt.Printf("Kamelet '%s' does not exist.\n", args[0])
+		fmt.Fprintf(cmd.OutOrStdout(), "Kamelet '%s' does not exist.\n", args[0])
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/describe_kit.go:
```diff
@@ -81,12 +81,12 @@ func (command *describeKitCommandOptions) run(cmd *cobra.Command, args []string)
 		}
 	} else {
-		fmt.Printf("IntegrationKit '%s' does not exist.\n", args[0])
+		fmt.Fprintf(cmd.OutOrStdout(), "IntegrationKit '%s' does not exist.\n", args[0])
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/kamelet_delete.go:
```diff
@@ -139,6 +139,6 @@ func (command *kameletDeleteCommandOptions) delete(name string) error {
 		return fmt.Errorf("error deleting kamelet \"%s\": %w", name, err)
 	}
-	fmt.Printf("kamelet \"%s\" has been deleted\n", name)
+	fmt.Fprintf(cmd.OutOrStdout(), "kamelet \"%s\" has been deleted\n", name)
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/kit_create.go:
```diff
@@ -175,22 +175,22 @@ func (command *kitCreateCommandOptions) run(_ *cobra.Command, args []string) err
 
 	if !existed {
-		fmt.Printf("integration kit \"%s\" created\n", kit.Name)
+		fmt.Fprintf(cmd.OutOrStdout(), "integration kit \"%s\" created\n", kit.Name)
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/kit_create.go:
```diff
@@ -175,22 +175,22 @@ func (command *kitCreateCommandOptions) run(_ *cobra.Command, args []string) err
+		fmt.Fprintf(cmd.OutOrStdout(), "integration kit \"%s\" created\n", kit.Name)
 	} else {
-		fmt.Printf("integration kit \"%s\" updated\n", kit.Name)
+		fmt.Fprintf(cmd.OutOrStdout(), "integration kit \"%s\" updated\n", kit.Name)
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/kit_delete.go:
```diff
@@ -144,7 +144,7 @@ func (command *kitDeleteCommandOptions) delete(name string) error {
 	}
 
-	fmt.Printf("integration kit \"%s\" has been deleted\n", kit.Name)
+	fmt.Fprintf(cmd.OutOrStdout(), "integration kit \"%s\" has been deleted\n", kit.Name)
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/log.go:
```diff
@@ -128,8 +128,8 @@ func (o *logCmdOptions) run(cmd *cobra.Command, args []string) error {
 			//
-			fmt.Printf("Integration '%s' is now running. Showing log ...\n", integrationID)
-			if err := k8slog.Print(o.Context, c, &integration, cmd.OutOrStdout()); err != nil {
+			fmt.Fprintf(cmd.OutOrStdout(), "Integration '%s' is now running. Showing log ...\n", integrationID)
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/modeline.go:
```diff
@@ -71,7 +71,7 @@ func NewKamelWithModelineCommand(ctx context.Context, osArgs []string) (*cobra.C
 	rootCmd, flags, err := createKamelWithModelineCommand(ctx, originalFlags)
 	if err != nil {
-		fmt.Printf("Error: %s\n", err.Error())
+		fmt.Fprintf(rootCmd.ErrOrStderr(), "Error: %s\n", err.Error())
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/uninstall.go:
```diff
@@ -209,71 +209,71 @@ func (o *uninstallCmdOptions) uninstallClusterWideResources(ctx context.Context,
 }
 
-func (o *uninstallCmdOptions) uninstallNamespaceResources(ctx context.Context, c client.Client) error {
+func (o *uninstallCmdOptions) uninstallNamespaceResources(cmd *cobra.Command, ctx context.Context, c client.Client) error {
```
The `ctx` argument should be declared in first position.

astefanutti commented on pkg/cmd/rebuild.go:
```diff
@@ -69,7 +69,7 @@ func (o *rebuildCmdOptions) rebuild(_ *cobra.Command, args []string) error {
 	}
 
-	fmt.Printf("%d integrations have been rebuilt\n", len(integrations))
+	fmt.Fprintf(cmd.OutOrStdout(), "%d integrations have been rebuilt\n", len(integrations))
```
`Fprintln` can be used here also.

astefanutti commented on pkg/cmd/run.go:
```diff
@@ -287,10 +287,10 @@ func (o *runCmdOptions) run(cmd *cobra.Command, args []string) error {
 				return
 			}
-			fmt.Printf("Run integration terminating\n")
+			fmt.Fprintf(cmd.OutOrStdout(), "Run integration terminating\n")
```
`Fprintln` can be used here also.

astefanutti reviewed (commented):
Thanks. There are a couple of leftovers for ``Fprintln`. It'll be good to go after that 👍🏼.

KuthumiPepple commented:
@astefanutti I have implemented the changes and fixed the lint errors. I also added a fix to a small error I noticed [here](https://github.com/apache/camel-k/blob/main/pkg/cmd/version.go#L176-L184). The print statements should use `aVersion` and `bVersion` not `a` and `b`. Anyway, kindly review and let me know if there are more things I need to do.

astefanutti reviewed (approved):
Thanks!

KuthumiPepple commented:
@astefanutti I have undone a change I made by accident please try running just the openshift-build check.
